### DISCUSSION TITLE
local resolver args null

### DISCUSSION THREAD

**Discussion by MarMun** [Akryum/vue-apollo#1061] [Q&A]
I'm trying to implement local resolvers with ` "apollo-client": "^2.6.10"` and `"vue-apollo": "^3.0.4"`.

Problem: Local resolver don't get `args` object (is always null) (in this case `args.id` is expected to be present).

The other parameters `parent`, `context`, `info` are all provided

How I try it:

```
// schema.js

const gql = require('graphql-tag')

export default gql`

  extend type Query {
    userTest(id: ID!): User
  }

  type UserTest {
    id: ID!
    name: String!
  }

`
```
```
// actions.js

import gql from 'graphql-tag'

const GET = gql`
  query userTest ($id: ID!) {
    userTest @client (
      id: $id
    ){
      id
      name
    }
  }
`
export default {
  GET
}
```

```
// resolver.js

export default {

  Query: {
    userTest (parent, args, context, info) {
      console.log(args) // null
      return {
        id: 'test',
        name: 'Martin',
        __typename: 'UserTest'
      }
    }
 }
}
```

Putting it together:

```
// main.js

import schema from './schema.js'
import resolver from './resolver.js'

.
.
.

// Create the apollo client
const apolloClient = new ApolloClient({
  link: links,
  cache,
  typeDefs: [schema],
  resolvers: [resolver]
})

// Create provider for apollo
const apolloProvider = new VueApollo({
  defaultClient: apolloClient
})

// Add apollo to Vue
Vue.use(VueApollo)

.
.
.
```

request the information:

```
// AppUser.vue

<script>

import actions from './actions.js'

export default {
  name: 'AppUser',
  components: { . . . },
  data () {
    return {
      id: 'UserTest_bbac9679-3342-4836-8500-999838b19b59'
    }
  },
  apollo: {
    userTest: {
      query: actions.GET,
      variables () {
        return {
          id: this.id
        }
      }
    }
  }
}
</script>
```

Any idea what I'm doing wrong?

**Comment by MarMun** [comment]
Important note: This refers to `@apollo/client/@3.2.1` (I switched to it to see if the problem mentioned above is maybe fixed in AC3)

Ok. I put my non existent apollo-client-internals knowledge to the test and stumbled over [this](https://github.com/apollographql/apollo-client/blob/main/src/utilities/graphql/storeUtils.ts#L237): 

```
// storeUtils.ts

export function argumentsObjectFromField(
  field: FieldNode | DirectiveNode,
  variables?: Record<string, any>,
): Object | null {
  if (field.arguments && field.arguments.length) {
    const argObj: Object = {};
    field.arguments.forEach(({ name, value }) =>
      valueToObjectRepresentation(argObj, name, value, variables),
    );
    return argObj;
  }
  return null;
}
```
This function returns null instead of the expected `args` object.

The `variables` object provided to this function looks like this:

`{id: "UserTest_bbac9679-3342-4836-8500-999838b19b59"}`

The `field` object provided to this function looks like this:
```
{
  "kind": "Field",
  "name": {
    "kind": "Name",
    "value": "userTest"
  },
  "arguments": [],
  "directives": [
    {
      "kind": "Directive",
      "name": {
        "kind": "Name",
        "value": "client"
      },
      "arguments": [
        {
          "kind": "Argument",
          "name": {
            "kind": "Name",
            "value": "id"
          },
          "value": {
            "kind": "Variable",
            "name": {
              "kind": "Name",
              "value": "id"
            }
          }
        }
      ]
    }
  ],
  "selectionSet": {
    "kind": "SelectionSet",
    "selections": [
      {
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "id"
        },
        "arguments": [],
        "directives": []
      },
      {
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "name"
        },
        "arguments": [],
        "directives": []
      },
      {
        "kind": "Field",
        "name": {
          "kind": "Name",
          "value": "__typename"
        }
      }
    ]
  }
}
```

The function expects `field.arguments` to be not empty, or it returns null.

Why `field.arguments` is empty is beyond my knowledge.

Maybe a incompatibility cause I use the unsupported apollo-client version 3 (although I have the same behavior as with apollo-client 2.6.1)?

**Comment by dmitry** [comment]
Local state resolver was deprecated since 3.x, it's better to use https://www.apollographql.com/docs/react/local-state/managing-state-with-field-policies. As I know vue-apollo is still not supporting 3.x version.
Little more on that: https://github.com/vuejs/vue-apollo/issues/1063

**Comment by MarMun** [comment]
Hey, thanks for your reply!

> Local state resolver was deprecated since 3.x,

That's true,but they still work and "deprecation and removal is still a ways off" ([source](https://github.com/apollographql/apollo-client/issues/7072#issuecomment-698978474)).

Plus, I have the same behavior with apollo-client 2.6.1 (args field is null)...
